### PR TITLE
add virtual com port utility to tools

### PR DESCRIPTION
**Pull request by rabryan**
- implements a serial terminal for communicating
with cdc and ftdi usb devices

- useful for debugging serial communications using
pyusb driver (especially on windows where switching
to serial driver from pyusb is a hassle)

### CODE REVIEW

**Comment by walac** [comment]
nit: make the code [PEP 8](https://www.python.org/dev/peps/pep-0008/) friend.

**Comment by rabryan** [comment]
Ok updated.

File: tools/vcp_terminal.py [hunk 1]
```diff
@@ -0,0 +1,648 @@
...
+        brates = [300, 600, 1200, 2400, 4800, 9600, 14400,
+                  19200, 28800, 38400, 57600, 115200, 230400]
+
+        for br in brates:
+            br_ftdi, value, index = ftdi_to_clkbits(br)
+            print("{:7} | {:7} | {:5.1%} | wVal {:4X}h | wIndex {:4X}h".format(
+                br, br_ftdi, (br - br_ftdi) * 1.0 / br, value, index))
+
+
+# vim:shiftwidth=2
```
Comment (walac) [inline tools/vcp_terminal.py]: nit: remove enforcement of vim config.

File: tools/vcp_terminal.py [hunk 2]
```diff
@@ -0,0 +1,648 @@
...
+    "GET_ENCAPSULATED_RESPONSE": 0x01,
+    "SET_COMM_FEATURE": 0x02,
+    "GET_COMM_FEATURE": 0x03,
+    "CLEAR_COMM_FEATURE": 0x04,
+    "SET_LINE_CODING": 0x20,
+    "GET_LINE_CODING": 0x21,
+    "SET_CONTROL_LINE_STATE": 0x22,
+    "SEND_BREAK": 0x23,   # wValue is break time
+}
+
```
Comment (walac) [inline tools/vcp_terminal.py]: nit: remove extra empty line

File: tools/vcp_terminal.py [hunk 3]
```diff
@@ -0,0 +1,648 @@
...
+            data_itf = data_itfs[0]
+            itf_num = data_itf.bInterfaceNumber
+        else:
+            data_itfs = list(
+                usb.util.find_descriptor(
+                    cfg,
+                    find_all=True,
+                    custom_match=lambda e: (
+                        e.bInterfaceClass == 0xA)))
+            if len(data_itfs) == 0:
```
Comment (walac) [inline tools/vcp_terminal.py]: s/`len(data_itfs) == 0`/`not data_itfs`/

File: tools/vcp_terminal.py [hunk 4]
```diff
@@ -0,0 +1,648 @@
...
+
+            data_itf = data_itfs[0]
+            cmd_itfs = list(
+                usb.util.find_descriptor(
+                    cfg,
+                    find_all=True,
+                    custom_match=lambda e: (
+                        e.bInterfaceClass == 0x2)))
+            itf_num = cmd_itfs[0].bInterfaceNumber
+            if(len(cmd_itfs) != len(data_itfs)):
```
Comment (walac) [inline tools/vcp_terminal.py]: nit: remove surrounding parenthesis

File: tools/vcp_terminal.py [hunk 5]
```diff
@@ -0,0 +1,648 @@
...
+        self._rxactive = False
+
+    def _read(self):
+        """ check ep for data, add it to queue and sleep for interval """
+        while self._rxactive:
+            try:
+                rv = self._ep_in.read(self._ep_in.wMaxPacketSize)
+                if self._isFTDI:
+                    status = rv[
+                        :2]       # FTDI prepends 2 flow control characters,
```
Comment (walac) [inline tools/vcp_terminal.py]: nit: lines 193 and 194 should be a single line.

File: tools/vcp_terminal.py [hunk 6]
```diff
@@ -0,0 +1,648 @@
...
+        while not self._rxqueue.empty():
+            rx.append(self._rxqueue.get())
+        return rx
+
+    def readText(self):
+        return "".join(chr(c) for c in self.readBytes())
+
+    def write(self, data):
+        ret = self._ep_out.write(data)
+        if(len(data) != ret):
```
Comment (walac) [inline tools/vcp_terminal.py]: nit: remove surround parenthesis

File: tools/vcp_terminal.py [hunk 7]
```diff
@@ -0,0 +1,648 @@
...
+                raise ValueError(
+                    "Invalid baudrates, nearest valid is {}".format(best))
+            self.baudrate = baudrate
+
+        if self._isFTDI:
+            self._setBaudFTDI(self.baudrate)
+            self._setLineCodeFTDI(
+                bits=self.databits,
+                stopbits=sbits[
+                    self.stopbits],
```
Comment (walac) [inline tools/vcp_terminal.py]: nit: lines 290 and 289 can be merge into a single line

File: tools/vcp_terminal.py [hunk 8]
```diff
@@ -0,0 +1,648 @@
...
+                parity=self.parity,
+                breaktype=0)
+        else:
+            linecode = [
+                self.baudrate & 0xff,
+                (self.baudrate >> 8) & 0xff,
+                (self.baudrate >> 16) & 0xff,
+                (self.baudrate >> 24) & 0xff,
+                sbits[
+                    self.stopbits],
```
Comment (walac) [inline tools/vcp_terminal.py]: nit: merge lines 299 and 300.

File: tools/vcp_terminal.py [hunk 9]
```diff
@@ -0,0 +1,648 @@
...
+        fileHandler.setFormatter(log_fmt)
+        log.addHandler(fileHandler)
+    return log
+
+
+def selectDevice():
+    devices = [d for d in usb.find(find_all=True)
+               if d.bDeviceClass in {0, 2, 0xff}]
+
+    if len(devices) == 0:
```
Comment (walac) [inline tools/vcp_terminal.py]: s/`len(devices) == 0`/`not devices`/

File: tools/vcp_terminal.py [hunk 10]
```diff
@@ -0,0 +1,648 @@
...
+            print(
+                "Please enter number between 0 and {}".format(
+                    len(devices) - 1))
+
+    # try:
+    d = devices[selection]
+    # except Exception as e:
+    #  log.error("Cannot open com port for {}".format(str(d)))
+    #  print(str(e))
+    #  exit()
```
Comment (walac) [inline tools/vcp_terminal.py]: nit: remove commented code.

File: tools/vcp_terminal.py [hunk 11]
```diff
@@ -0,0 +1,648 @@
...
+
+
+def runTerminal(d):
+    log.info("Beginning a terminal run")
+    p = ComPort(d)
+    p.setControlLineState(True, True)
+    p.setLineCoding()
+    # try:
+    # except usb.USBError as e:
+    #  log.error("USBErr: {}".format(e))
```
Comment (walac) [inline tools/vcp_terminal.py]: nit: remove commented code.

**Comment by walac** [comment]
PEP 8 [recommends snake_case for function and variable names](https://www.python.org/dev/peps/pep-0008/#naming-conventions), but as this is an external tool, it is up to you if you want to follow this convention or not.

Other than that, there are some nits but the code looks good in general.

File: tools/vcp_terminal.py (hunk 3 above)
Comment (rabryan) [inline tools/vcp_terminal.py]: IMO using len() makes the intent a little clearer, but I'll update this to conform to the project style.

**Comment by rabryan** [comment]
Ok.   I updated with changes addressing those issues.  Also, added some better exception handling.

**Comment by walac** [comment]
Could you please squash the three commits before merging?

**Comment by rabryan** [comment]
Ok

**Comment by walac** [comment]
Thanks.

**Pull request merged**